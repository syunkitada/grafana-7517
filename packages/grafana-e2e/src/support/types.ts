import { Selector } from './selector';
import { fromBaseUrl } from './url';
import { e2e } from '../index';
import { SelectorFunction, VisitFunction } from '../noTypeCheck';

export type Selectors = Record<string, string | Function>;
export type SelectorFunctions<S> = { [P in keyof S]: SelectorFunction };

export type Page<S> = SelectorFunctions<S> & {
  selectors: S;
  visit: VisitFunction;
};
export interface PageFactoryArgs<S> {
  selectors: S;
  url?: string | Function;
}

export const pageFactory = <S extends Selectors>({ url, selectors }: PageFactoryArgs<S>): Page<S> => {
  const visit = (args?: string) => {
    if (!url) {
      return e2e().visit('');
    }

    let parsedUrl = '';
    if (typeof url === 'string') {
      parsedUrl = fromBaseUrl(url);
    }

    if (typeof url === 'function' && args) {
      parsedUrl = fromBaseUrl(url(args));
    }

    e2e().logToConsole('Visiting', parsedUrl);
    return e2e().visit(parsedUrl);
  };
  const pageObjects: SelectorFunctions<S> = {} as SelectorFunctions<S>;
  const keys = Object.keys(selectors);

  keys.forEach(key => {
    const value = selectors[key];
    if (typeof value === 'string') {
      // @ts-ignore
      pageObjects[key] = () => {
        e2e().logToConsole('Retrieving Selector:', value);
        return e2e().get(Selector.fromAriaLabel(value));
      };
    }
    if (typeof value === 'function') {
      // @ts-ignore
      pageObjects[key] = (text?: string) => {
        if (!text) {
          const selector = value();
          e2e().logToConsole('Retrieving Selector:', selector);
          return e2e().get(selector);
        }
        const selector = value(text);
        e2e().logToConsole('Retrieving Selector:', selector);
        return e2e().get(Selector.fromAriaLabel(selector));
      };
    }
  });

  return {
    visit,
    ...pageObjects,
    selectors,
  };
};

type Component<S> = Omit<Page<S>, 'visit'>;
type ComponentFactoryArgs<S> = Omit<PageFactoryArgs<S>, 'url'>;

export const componentFactory = <S extends Selectors>(args: ComponentFactoryArgs<S>): Component<S> => {
  const { visit, ...rest } = pageFactory(args);
  return rest;
};
