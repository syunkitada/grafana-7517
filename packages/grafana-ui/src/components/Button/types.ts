import { GrafanaTheme } from '@grafana/data';

export type ButtonVariant = 'primary' | 'secondary' | 'danger' | 'inverse' | 'transparent' | 'destructive' | 'link';

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg';

export interface StyleDeps {
  theme: GrafanaTheme;
  size: ButtonSize;
  variant: ButtonVariant;
}

export interface ButtonStyles {
  button: string;
  buttonWithIcon: string;
  iconButton: string;
  iconWrap: string;
  icon?: string;
}
