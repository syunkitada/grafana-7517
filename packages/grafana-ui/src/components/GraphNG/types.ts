import { DataFrameFieldIndex } from '@grafana/data';

/**
 * Mode to describe if a legend is isolated/selected or being appended to an existing
 * series selection.
 * @public
 */
export enum GraphNGLegendEventMode {
  ToggleSelection = 'select',
  AppendToSelection = 'append',
}

/**
 * Event being triggered when the user interact with the Graph legend.
 * @public
 */
export interface GraphNGLegendEvent {
  fieldIndex: DataFrameFieldIndex;
  mode: GraphNGLegendEventMode;
}
