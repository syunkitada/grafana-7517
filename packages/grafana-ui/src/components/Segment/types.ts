import { ReactElement } from 'react';
import { SelectableValue } from '@grafana/data';

export interface SegmentProps<T> {
  onChange: (item: SelectableValue<T>) => void;
  value?: SelectableValue<T>;
  Component?: ReactElement;
  className?: string;
  allowCustomValue?: boolean;
}
