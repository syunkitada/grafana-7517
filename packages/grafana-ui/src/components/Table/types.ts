import { TextAlignProperty } from 'csstype';
import { ComponentType } from 'react';
import { Field } from '@grafana/data';
import { TableStyles } from './styles';

export interface TableFieldOptions {
  width: number;
  align: FieldTextAlignment;
  displayMode: TableCellDisplayMode;
}

export enum TableCellDisplayMode {
  Auto = 'auto',
  ColorText = 'color-text',
  ColorBackground = 'color-background',
  GradientGauge = 'gradient-gauge',
  LcdGauge = 'lcd-gauge',
}

export type FieldTextAlignment = 'auto' | 'left' | 'right' | 'center';

export interface TableColumn {
  // React table props
  Header: string;
  accessor: string | Function;
  Cell: ComponentType<ReactTableCellProps>;
  // Grafana additions
  field: Field;
  width: number;
  textAlign: TextAlignProperty;
}

export interface TableRow {
  [x: string]: any;
}

export interface ReactTableCellProps {
  cell: ReactTableCell;
  column: TableColumn;
  tableStyles: TableStyles;
}

export interface ReactTableCell {
  value: any;
}
