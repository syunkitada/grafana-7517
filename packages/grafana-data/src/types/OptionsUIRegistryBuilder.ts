import { ComponentType } from 'react';
import { RegistryItem, Registry } from '../utils/Registry';
import { NumberFieldConfigSettings, SelectFieldConfigSettings, StringFieldConfigSettings } from '../field';

/**
 * Option editor registry item
 */
export interface OptionsEditorItem<TOptions, TSettings, TEditorProps, TValue> extends RegistryItem {
  path: (keyof TOptions & string) | string;
  editor: ComponentType<TEditorProps>;
  settings?: TSettings;
  defaultValue?: TValue;
}

/**
 * Configuration of option editor registry item
 */
interface OptionEditorConfig<TOptions, TSettings, TValue = any> {
  id: keyof TOptions & string;
  name: string;
  description: string;
  settings?: TSettings;
  defaultValue?: TValue;
}

/**
 * Describes an API for option editors UI builder
 */
export interface OptionsUIRegistryBuilderAPI<
  TOptions,
  TEditorProps,
  T extends OptionsEditorItem<TOptions, any, TEditorProps, any>
> {
  addNumberInput?<TSettings extends NumberFieldConfigSettings = NumberFieldConfigSettings>(
    config: OptionEditorConfig<TOptions, TSettings, number>
  ): this;

  addTextInput?<TSettings extends StringFieldConfigSettings = StringFieldConfigSettings>(
    config: OptionEditorConfig<TOptions, TSettings, string>
  ): this;

  addSelect?<TOption, TSettings extends SelectFieldConfigSettings<TOption>>(
    config: OptionEditorConfig<TOptions, TSettings, TOption>
  ): this;

  addRadio?<TOption, TSettings extends SelectFieldConfigSettings<TOption> = SelectFieldConfigSettings<TOption>>(
    config: OptionEditorConfig<TOptions, TSettings, TOption>
  ): this;

  addBooleanSwitch?<TSettings = any>(config: OptionEditorConfig<TOptions, TSettings, boolean>): this;

  addUnitPicker?<TSettings = any>(config: OptionEditorConfig<TOptions, TSettings, string>): this;

  addColorPicker?<TSettings = any>(config: OptionEditorConfig<TOptions, TSettings, string>): this;

  /**
   * Enables custom editor definition
   * @param config
   */
  addCustomEditor<TSettings, TValue>(config: OptionsEditorItem<TOptions, TSettings, TEditorProps, TValue>): this;

  /**
   * Returns registry of option editors
   */
  getRegistry: () => Registry<T>;
}

export abstract class OptionsUIRegistryBuilder<
  TOptions,
  TEditorProps,
  T extends OptionsEditorItem<TOptions, any, TEditorProps, any>
> implements OptionsUIRegistryBuilderAPI<TOptions, TEditorProps, T> {
  private properties: T[] = [];

  addCustomEditor<TSettings, TValue>(config: T & OptionsEditorItem<TOptions, TSettings, TEditorProps, TValue>): this {
    this.properties.push(config);
    return this;
  }

  getRegistry() {
    return new Registry(() => {
      return this.properties;
    });
  }
}
