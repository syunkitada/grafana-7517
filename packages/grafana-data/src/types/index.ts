export * from './data';
export * from './logs';
export * from './navModel';
export * from './time';
export * from './utils';
