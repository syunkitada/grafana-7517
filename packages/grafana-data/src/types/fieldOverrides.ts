import { ComponentType } from 'react';
import {
  MatcherConfig,
  FieldConfig,
  Field,
  DataFrame,
  VariableSuggestionsScope,
  VariableSuggestion,
  GrafanaTheme,
  TimeZone,
} from '../types';
import { Registry } from '../utils';
import { InterpolateFunction } from './panel';
import { StandardEditorProps } from '../field';
import { OptionsEditorItem } from './OptionsUIRegistryBuilder';

export interface DynamicConfigValue {
  prop: string;
  value?: any;
  custom?: boolean;
}

export interface ConfigOverrideRule {
  matcher: MatcherConfig;
  properties: DynamicConfigValue[];
}

export interface FieldConfigSource<TOptions extends object = any> {
  // Defatuls applied to all numeric fields
  defaults: FieldConfig<TOptions>;

  // Rules to override individual values
  overrides: ConfigOverrideRule[];
}

export interface FieldOverrideContext {
  field?: Field;
  dataFrameIndex?: number; // The index for the selected field frame
  data: DataFrame[]; // All results
  replaceVariables?: InterpolateFunction;
  getSuggestions?: (scope?: VariableSuggestionsScope) => VariableSuggestion[];
}

export interface FieldConfigEditorProps<TValue, TSettings>
  extends Omit<StandardEditorProps<TValue, TSettings>, 'item'> {
  item: FieldPropertyEditorItem<TValue, TSettings>; // The property info
  value: TValue;
  context: FieldOverrideContext;
  onChange: (value?: TValue) => void;
}

export interface FieldOverrideEditorProps<TValue, TSettings> extends Omit<StandardEditorProps<TValue>, 'item'> {
  item: FieldPropertyEditorItem<TValue, TSettings>;
  context: FieldOverrideContext;
}

export interface FieldConfigEditorConfig<TOptions, TSettings = any, TValue = any> {
  id: (keyof TOptions & string) | string;
  name: string;
  description: string;
  settings?: TSettings;
  shouldApply?: (field: Field) => boolean;
  defaultValue?: TValue;
}

export interface FieldPropertyEditorItem<TOptions = any, TValue = any, TSettings extends {} = any>
  extends OptionsEditorItem<TOptions, TSettings, FieldConfigEditorProps<TValue, TSettings>, TValue> {
  // An editor that can be filled in with context info (template variables etc)
  override: ComponentType<FieldOverrideEditorProps<TValue, TSettings>>;

  // Convert the override value to a well typed value
  process: (value: any, context: FieldOverrideContext, settings?: TSettings) => TValue;

  // Checks if field should be processed
  shouldApply: (field: Field) => boolean;
}

export type FieldConfigEditorRegistry = Registry<FieldPropertyEditorItem>;

export interface ApplyFieldOverrideOptions {
  data?: DataFrame[];
  fieldOptions: FieldConfigSource;
  replaceVariables: InterpolateFunction;
  theme: GrafanaTheme;
  timeZone?: TimeZone;
  autoMinMax?: boolean;
  standard?: FieldConfigEditorRegistry;
  custom?: FieldConfigEditorRegistry;
}

export enum StandardFieldConfigProperties {
  Unit = 'unit',
  Min = 'min',
  Max = 'max',
  Decimals = 'decimals',
  Title = 'title',
  NoValue = 'noValue',
  Thresholds = 'thresholds',
  Mappings = 'mappings',
  Links = 'links',
  Color = 'color',
}
