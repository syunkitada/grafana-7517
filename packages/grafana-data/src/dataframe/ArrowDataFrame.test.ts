import fs from 'fs';
import path from 'path';

import { grafanaDataFrameToArrowTable, arrowTableToDataFrame } from './ArrowDataFrame';
import { toDataFrameDTO, toDataFrame } from './processDataFrame';
import { FieldType } from '../types';
import { Table } from 'apache-arrow';

describe('Read/Write arrow Table to DataFrame', () => {
  test('should parse output with dataframe', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      meta: {
        aaa: 'xyz',
        anything: 'xxx',
      },
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'value', config: { min: 0, max: 50, unit: 'somthing' }, type: FieldType.number, values: [1, 2, 3] },
        { name: 'str', config: {}, type: FieldType.string, values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should support duplicate field names', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'a', values: [1, 2, 3] },
        { name: 'a', values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should read all types', () => {
    const fullpath = path.resolve(__dirname, './__snapshots__/all_types.golden.arrow');
    const arrow = fs.readFileSync(fullpath);
    const table = Table.from([arrow]);
    const frame = arrowTableToDataFrame(table);
    expect(toDataFrameDTO(frame)).toMatchSnapshot();
  });
});
