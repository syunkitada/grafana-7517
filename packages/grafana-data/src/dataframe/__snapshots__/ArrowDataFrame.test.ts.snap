// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GEL Utils should parse output with dataframe 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "Time",
        "type": "time",
        "values": Array [
          1569334575000,
          1569334580000,
          1569334585000,
          1569334590000,
          1569334595000,
          1569334600000,
          1569334605000,
          1569334610000,
          1569334615000,
          1569334620000,
          1569334625000,
          1569334630000,
          1569334635000,
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "",
        "type": "number",
        "values": Array [
          3,
          3,
          3,
          5,
          5,
          5,
          3,
          3,
          3,
          5,
          5,
          5,
          3,
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": "GC",
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "Time",
        "type": "time",
        "values": Array [
          1569334575000,
          1569334580000,
          1569334585000,
          1569334590000,
          1569334595000,
          1569334600000,
          1569334605000,
          1569334610000,
          1569334615000,
          1569334620000,
          1569334625000,
          1569334630000,
          1569334635000,
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "GB-series",
        "type": "number",
        "values": Array [
          0,
          0,
          0,
          2,
          2,
          2,
          0,
          0,
          0,
          2,
          2,
          2,
          0,
        ],
      },
    ],
    "meta": undefined,
    "name": undefined,
    "refId": "GB",
  },
]
`;
